[TextServer] Implement ICU/UAX#31 based `is_valid_identifier` function.

bruvzg opened the pull request:
Partially implements https://github.com/godotengine/godot-proposals/issues/916

Implements `TextServer::is_valid_identifier` function, but do not change any of the GDScript related code.

TextServerAdvnaced version use following rules - a valid identifier must:
  - Conform to normalization form C.
  - Begin with a Unicode character of class XID_Start or `"_"`.
  - May contain Unicode characters of class XID_Continue in the other positions.
  - Use UAX#31 recommended scripts only (mixed scripts are allowed).

Fallback version uses `String::is_valid_identifier` which implements following rules - a valid identifier must:
  - Begin with a Basic Latin letter character or `"_"`.
  - May contain Basic Latin letter characters, digits or `"_"` in the other positions.

bruvzg committed:
[TextServer] Implement ICU/UAX 31 based `is_valid_identifier` function.

akien-mga commented on doc/classes/TextServer.xml:
```diff
@@ -798,6 +798,21 @@
+			<argument index="0" name="string" type="String" />
+			<description>
+				Returns [code]true[/code] is [code]string[/code] is a valid identifier.
+				If text server supports [code]FEATURE_BIDI_LAYOUT[/code] feature, a valid identifier must:
```
```suggestion
				If the text server supports the [constant FEATURE_BIDI_LAYOUT] feature, a valid identifier must:
```

akien-mga commented on doc/classes/TextServer.xml:
```diff
@@ -798,6 +798,21 @@
+				- Begin with a Unicode character of class XID_Start or [code]"_"[/code].
+				- May contain Unicode characters of class XID_Continue in the other positions.
+				- Use UAX #31 recommended scripts only (mixed scripts are allowed).
+				If [code]FEATURE_BIDI_LAYOUT[/code] feature is not supported, a valid identifier must:
```
```suggestion
				If the [constant FEATURE_BIDI_LAYOUT] feature is not supported, a valid identifier must:
```

timothyqiu commented:
IINM, `TextServer` is for "rendering" fonts and text layouts. I think putting `is_valid_identifier` here might cause some trouble. e.g. A plugin using Unicode identifiers internally might not work in projects using an incompatible text driver.

A valid GDScript should be always valid no matter using advanced text server or not.

reduz commented:
Is this missing anything?

bruvzg commented:
> Is this missing anything?

No, it's ready, but it and #54905 are pretty useless by itself, until Unicode support is implemented in the GDScript parser (not sure what is the final decision about it). Also, since it a part of optional TextServer, using it would make GDScript parser dependent on a specific TextServer implementation.

reduz commented:
Code looks good to me, but left to assess that this approach is desired.

reduz commented:
@bruvzg I think we discussed at some point that for text servers that are not advanced, we could have a more hacky implementation, like using a table of valid/invalid characters that may have false positives but that since this is supposed to be checked properly in the editor, it should still be allowed.

bruvzg commented:
> we could have a more hacky implementation, like using a table of valid/invalid characters that may have false positives but that since this is supposed to be checked properly in the editor, it should still be allowed.

I guess we can have `XID_Start` and `XID_Continue` checks by using simple lookup tables, and the rest (script check and normalization) is not that important. And spoofing checks (https://github.com/godotengine/godot/pull/54905) are completly optional, it's probably enough to document that it's not supported with fallback text server.

reduz commented:
I think the logic we should implement is the following for GDScript, Expression and VariantParser

1) Have a function that lets you know that some character is a valid identifier character, this includes control or whathever characters, we can a big hardcoded table with ranges for this that you can generate from ICU and this is harcoded in Godot, outside TextServer.
2) For the text servers that don't have ICU or text data needed to actually determine what is a valid identifier, have a function that uses the above table and lets you know that its valid or not (even if there are false positives). For text servers that support it, do it properly.

The reason this is needed is that GDScript, Expression, VariantParser, etc. use char by char parsing to determine identifiers. So the logic should be that it first gets the whole String and then it should determine that something is a valid identifier or not.

bruvzg commented:
Added `is_unicode_identifier_start` and `is_unicode_identifier_continue` (in the `char_utils.h`) functions and updated fallback text server to use it.

`TextServerAdvnaced` version use following rules - a valid identifier must:

- Conform to normalization form C.
- Begin with a Unicode character of class XID_Start or "_".
- May contain Unicode characters of class XID_Continue in the other positions.
- May contain ZWNJ/ZWJ in the special cases, see [Layout and Format Control Characters](https://www.unicode.org/reports/tr31/#Layout_and_Format_Control_Characters).
- Use UAX#31 recommended scripts only (mixed scripts are allowed).

Fallback implements the following rules - a valid identifier must:

- Begin with a Unicode character of class XID_Start or "_".
- May contain Unicode characters of class XID_Continue in the other positions.

akien-mga commented on core/string/char_utils.h:
```diff
@@ -33,6 +33,1444 @@
 
 #include "core/typedefs.h"
 
+struct CharRange {
+	char32_t start;
+	char32_t end;
+};
+
+static CharRange xid_start[] = {
+	{ 0x41, 0x5a },
+	{ 0x5f, 0x5f },
+	{ 0x61, 0x7a },
+	{ 0xaa, 0xaa },
+	{ 0xb5, 0xb5 },
+	{ 0xba, 0xba },
+	{ 0xc0, 0xd6 },
+	{ 0xd8, 0xf6 },
```
Maybe these could go to a `core/string/char_range.inc` to reduce the clutter in `char_utils.h` (which would be included here for the two new methods that use this data)? Having so much data in a header makes the actual logic harder to consult IMO.

bruvzg replied:
Moved ranges to a separate file.

akien-mga commented on doc/classes/TextServer.xml:
```diff
@@ -944,6 +944,21 @@
+			<description>
+				Returns [code]true[/code] is [code]string[/code] is a valid identifier.
+				If the text server supports the [constant FEATURE_UNICODE_IDENTIFIERS] feature, a valid identifier must:
+				- Conform to normalization form C.
```
```suggestion
				- Conform to normalization from C.
```
? Or is "form C" a type of form?

bruvzg replied:
It's "form C", see https://www.unicode.org/reports/tr15/#Norm_Forms

akien-mga commented on core/string/char_utils.h:
```diff
@@ -31,7 +31,25 @@
 #define CHAR_UTILS_H
 
-#include "core/typedefs.h"
+#include "char_range.h"
```
Not too important but we usually name such data-only includes with `.inc` extension.

akien-mga commented:
Thanks!
